autorecsys/pipeline/preprocessor.py

```diff
@@ -78,6 +103,216 @@ def preprocessing(self, **kwargs):
+
+        # Step 3: Index categories.
+        for i, categ_l in enumerate(categ_ll):
+            for j, c in enumerate(categ_l):
+                if c in categ_index_dict[i]:
+                    categ_ll[i][j] = categ_index_dict[i][c]  # index in-place
+                else:  # index filtered categories as the end index
+                    categ_ll[i][j] = len(categ_index_dict[i])
+        del categ_index_dict  # release memory
+        gc.collect()
+
+        # Step 4: Format data and obtain hash statistics.
+        array_data = np.concatenate((np.asarray(label_ll).T, np.asarray(numer_ll).T, np.asarray(categ_ll).T), axis=1)
+        del label_ll  # release memory
+        del numer_ll
+        del categ_ll
+        gc.collect()
+
+        self.pd_data = pd.DataFrame(array_data, columns=self.used_columns_names)
+        self.hash_sizes = [self.pd_data[categ_name].nunique() for categ_name in self.categ_names]
+        del array_data  # release memory
+        gc.collect()
+
+        for col_name, dtype in self.dtype_dict.items():
+            self.pd_data[col_name] = self.pd_data[col_name].astype(dtype)
+        # Save transformed data.
+        save_pickle(self.transformed_data_path, self.pd_data)
+
+    def scale_numerical_data(self):
+
+        # TODO: Designated transformation functions to specified place.
+        def scale_by_natural_log(num):
+            # TODO: 1) Explain why the conditional statement makes exception for numbers like 1, where 1>ln(1)**2
+            if num > 2:
+                num = math.log(float(num))**2
+            return num
+
+        for numer_name in self.numer_names:
+            self.pd_data[numer_name] = self.pd_data[numer_name].map(scale_by_natural_log)
+
+    def split_data(self, X, y, train_size, validate_size, random_state=42):
+        """
+
+        :param X: numpy array
+        :param y: numpy array
+        :param train_size:
+        :param validate_size:
+        :param random_state:
+        :return:
+        """
+        # Step 1: Obtain shuffled data indices.
+        data_size = X.shape[0]
+        np.random.seed(random_state)
+        shuffled_indices = np.random.permutation(range(data_size))
+
+        # Step 2: Determine data split positions.
+        train_end = int(train_size * data_size)
+        validate_end = train_end + int(validate_size * data_size)
+
+        # Step 3: Split shuffled data.
+        train_X = X[shuffled_indices][:train_end]
+        train_y = y[shuffled_indices][:train_end]
+        validate_X = X[shuffled_indices][train_end:validate_end]
+        validate_y = y[shuffled_indices][train_end:validate_end]
+        test_X = X[shuffled_indices][validate_end:]
+        test_y = y[shuffled_indices][validate_end:]
+
+        return train_X, validate_X, test_X, train_y, validate_y, test_y
+
+    def preprocessing(self, train_size, validate_size, random_state=42):
+        self.X = self.pd_data.iloc[:, 1:].values
```

Comment:
1. Please call the load_data & data split function here in the preprocessing so the user can not only directly use one function to do it all, but also enjoy the flexibility to do them separately.

2. Since it will do both load and preprocessing, I don't have a good name for it, you can think of it or we can keep it for now.

Reply:
Done, and let's keep the name for now.